demoapp/demo_app.py

```diff
@@ -58,224 +58,430 @@ def get_mgl_sections_file() -> pd.DataFrame:
+                section_name = names_df[(names_df["Chapter_Number"] == chap) & (names_df["Section_Number"] == sec)]['Section Name'].values[0]
+                names.append({chap: chapter_name, sec: section_name})
+            except Exception as e:
+                print("Missing Chapter or Section Name")
+                continue
+
+        return names
+
+def get_MGL_for_bill(df:pd.DataFrame, bill_number) -> str:
+    """
+    Outputs MGL section text from the dataframe referenced in the selected bill
+    Args:
+        df (pd.DataFrame): DataFrame containing bill and MGL texts
+        bill_number: str:Bill number of the selected bill
 
-def generate_response(bill_number: str, text:str, category:str) -> str:
-    """Function to generate response"""
+    Returns:
+        str: _description_
+    """
+    mgl_ref = df.loc[df['BillNumber']== bill_number, 'Combined_MGL'].values[0]
+    return mgl_ref
 
-    API_KEY = st.session_state["OPENAI_API_KEY"]
-    os.environ['OPENAI_API_KEY'] = API_KEY
-    # loader = TextLoader("demoapp/extracted_mgl.txt").load()
-    # bills_12_with_mgl.loc[bills_12_with_mgl['BillNumber'] == "H489", 'combined_MGL_y']
-    # grab MGL section based on selected bill number
-    mgl_ref = df.loc[df['BillNumber']== bill_number, 'Combined_MGL']
+def generate_response_large_documents(bill_title: str, bill_text: str, mgl_ref: str, mgl_names: List[Dict[str, str]], llm) -> str:
+    """
+    Generates a detailed response for large documents related to MA bills.
     
-    mgl_ref = str(mgl_ref.values[0])
+    Args:
+        bill_title (str): Title of the bill.
+        bill_text (str): Text content of the bill.
+        mgl_ref (Union[str, None]): Reference Mass General Law text, if any.
+        mgl_names (List[Dict[str, str]]): List of Mass General Law chapter and section names.
+        llm: Language model for generating responses.
     
-    text_splitter = CharacterTextSplitter(chunk_size=4000, chunk_overlap=0)
-    if mgl_ref is not None:
-        #splits MGL text into chunks of sizes specified in the previous step
-        documents = [Document(page_content=x) for x in text_splitter.split_text(mgl_ref)]
-    else:
-        documents = ""
-        
-    # text_splitter = CharacterTextSplitter(chunk_size=4000, chunk_overlap=0)
-    # documents = text_splitter.split_documents(loader)
+    Returns:
+        str: Detailed response generated by the language model.
+    """
+    #Splitting the text and using vector embeddings when document size is > 90000 words
     
+    #Context size for 'gpt-4-1106-preview' is 128K; Need to break up the documents and use vector embenddings when the word length of the documents is large
+    #Chose 90K cut off for large docs because: 1 token = 0.75 words --> 90000 words = ~120K words + leaving  room for response
+
+    text_splitter = CharacterTextSplitter(chunk_size=90000, chunk_overlap=1000)
```

Comment:
It's probably inefficient to etsablish a unique vector store for each prompt. But if it's quick enough to do in practice, this is probably fine.

Reply:
yeah, it shouldn't be splitting the text into more than 2 chunks. Also, it would only affect about 5% of the bills, the other ones won't be using vector stores.